### DISCUSSION TITLE
Crackle at start and end of audio received from HTTP Get Request with I2SStream PAM8304

### DISCUSSION THREAD

**Discussion by soham2560** [pschatzmann/arduino-audio-tools#1639] [Q&A] [answered]
#### Brief
I am currently trying to 
- use INMP441 with I2SStream to stream audio to a server through Websockets
- receive audio as a response from the server through a GET request that is copied to AnalogAudioStream PAM8304. 
- The microphone and speaker functionality are supposed to work one after the other (i.e. only one at a time). The speaker audio can start abruptly and end abruptly too.

#### Problems
- The speaker audio unfortunately has a crackle at both the start and end of its functioning. I would like to have a clear cut off. I have attached recorded audio of speaker at the bottom.

#### HTTP and Speaker Config
```bash
WiFiClient client;
HttpRequest http(client);
AnalogAudioStream audioOutput;
EncodedAudioStream decoder(&audioOutput, new MP3DecoderHelix()); 
StreamCopy copier(decoder, client);
```
```bash
auto cfg = audioOutput.defaultConfig(TX_MODE);
audioOutput.begin(cfg);
decoder.begin();
```
#### Code Snippet (runs in loop())
```bash
void loop()
{
  boot_button.loop();
  main_button.loop();
  webSocket.loop();

  switch (CURR_STATE)
  {
    case IDLE:
    {
      if(main_button.wasPressed())
      {
        switch(main_button.read())
        {
          case single_click:
          {
            CURR_STATE = USER_TALK;
          }
          break;
        }
      }
      break;
    }
    case USER_TALK:
    {
      Serial.println("USER_TALK");
      size_t bytesRead = conv.readBytes(audioData, BUFFER_SIZE);
      if (bytesRead > 0) {
        volumeScaler.convert(audioData, bytesRead);
        webSocket.sendBIN(audioData, bytesRead);
      }

      disconnect_handler();
      break;
    }
    case POLL_SERVER:
    {
      Serial.println("POLL_SERVER");
      while(!http.processBegin(GET, url, nullptr, -1))
      {
        delay(20);
        disconnect_handler();
      }
      while(!copier.available());
      CURR_STATE = SERVER_TALK;
      break;
    }
    case SERVER_TALK:
    {
      Serial.println("SERVER_TALK");
      if(!copier.copy())
      {
        CURR_STATE = USER_TALK;
      }

      disconnect_handler();
      break;
    }
  }
}
```
Note: The switch between the cases is also handled externally. A websocket message from the server once received causes the switch.

I get crackle at the start of audio and also whenever the copier ends in SERVER_TALK (due to copier.copy()! returning false or even the switch case changing).

https://github.com/user-attachments/assets/1251e357-41ab-4d19-a530-2171a946e2a0

Note the single crackle at the start, and the multiple at the end.

Originial audio file is [here](https://drive.google.com/file/d/1_AHwtJ4wOfwNXTD0j04qP0n4W5Gzmans/view?usp=sharing)

**Comment by pschatzmann** [comment, marked as answer]
Just read the Interactive Audio Wiki!
